File: ge-develop/examples/es/optional_input/python/src/make_batchnorm_graph.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
# ----------------------------------------------------------------------------
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of 
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, 
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ----------------------------------------------------------------------------

import os
import sys
import numpy as np

from ge.es.graph_builder import GraphBuilder, TensorHolder
from ge.graph import Tensor
from ge.graph.types import DataType, Format
from ge.graph import Graph, DumpFormat
from ge.ge_global import GeApi
from ge.session import Session
from ge.es.all import BatchNorm


def build_batch_norm_graph():
    # 1、创建图构建器
    builder = GraphBuilder("MakeBatchNormGarph")
    # 2、创建图输入节点 
    input_tensor_holder = builder.create_input(
        index=0,
        name="input",
        data_type=DataType.DT_FLOAT,
        format=Format.FORMAT_NCHW,
        shape=[1, 3, 1, 2]
    )
    mean = builder.create_input(
        index=1,
        name="mean",
        data_type=DataType.DT_FLOAT,
        shape=[3]
    )
    variance = builder.create_input(
        index=2,
        name="variance",
        data_type=DataType.DT_FLOAT,
        shape=[3]
    )
    scale = builder.create_const_float([1.0, 1.0, 1.0], shape=[3])
    offset = builder.create_const_float([0.0, 0.0, 0.0], shape=[3])
    batchNorm_tensor_holder = BatchNorm(input_tensor_holder, scale, offset, mean, variance,
                                        epsilon=1e-4, data_format="NCHW", is_training=False)
    # 3、设置图输出节点
    builder.set_graph_output(batchNorm_tensor_holder.y, 0)
    # 4、构建图
    return builder.build_and_reset()


def dump_batch_norm_graph(graph):
    graph.dump_to_file(format=DumpFormat.kOnnx, suffix="make_batch_norm_graph")


def run_graph(graph, device_id="0") -> int:
    # 1. 初始化GE环境
    config = {
        "ge.exec.deviceId": str(device_id),
        "ge.graphRunMode": "0"
    }

    ge_api = GeApi()
    ret = ge_api.ge_initialize(config)
    if ret != 0:
        print(f"[Error] GE初始化失败，返回码: {ret}")
        return ret

    print(f"[Info] GE环境初始化成功 (Device ID: {device_id})")

    try:
        # 2. 创建Session
        session = Session()

        # 3. 添加图到Session
        graph_id = 1
        ret = session.add_graph(graph_id, graph)
        if ret != 0:
            print(f"[Error] 添加图失败，返回码: {ret}")
            return ret
        print(f"[Info] 图已添加到Session (Graph ID: {graph_id})")

        # 4. 准备输入数据
        input_data = np.array([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], dtype=np.float32)
        mean_data = np.array([2.0, 4.0, 6.0], dtype=np.float32)
        variance_data = np.array([0.5, 0.5, 0.5], dtype=np.float32)

        input_tensor = Tensor(
            input_data.tolist(),
            None,
            DataType.DT_FLOAT,
            Format.FORMAT_NCHW,
            [1, 3, 1, 2]
        )
        mean_tensor = Tensor(
            mean_data.tolist(),
            None,
            DataType.DT_FLOAT,
            Format.FORMAT_ND,
            [3]
        )
        variance_tensor = Tensor(
            variance_data.tolist(),
            None,
            DataType.DT_FLOAT,
            Format.FORMAT_ND,
            [3]
        )

        inputs = [input_tensor, mean_tensor, variance_tensor]
        print(f"[Info] 输入数据已准备，共{len(inputs)}个输入tensor (input, mean, variance)")

        # 5. 运行图
        ret = session.run_graph(graph_id, inputs)
        print("[Info] 图运行成功！")
        for idx, tensor in enumerate(ret, start=1):
            print(f"Tensor{idx}详情：{tensor}")
        return 0

    except Exception as e:
        print(f"[Error] 执行过程中出错: {e}")
        import traceback
        traceback.print_exc()
        return -1

    finally:
        # 6. 清理GE环境
        print("[Info] 清理GE环境...")
        ge_api.ge_finalize()
        print("[Success] GE环境已清理")


if __name__ == "__main__":
    # 构建图
    graph = build_batch_norm_graph()

    # 先dump图（生成pbtxt文件用于可视化）
    dump_batch_norm_graph(graph)

    # 运行图（从命令行参数获取device_id，默认为"0"）
    run_graph(graph)

File: ge-develop/examples/es/optional_input/cpp/src/make_batchnorm_graph.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of 
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, 
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */
#include "es_showcase.h"
#include "es_BatchNorm.h"
#include "utils.h"
#include <memory>
#include <vector>
#include <iostream>
#include "ge/ge_api.h"

using namespace ge;
using namespace ge::es;
namespace {
es::EsTensorHolder MakeBatchNormGraph(es::EsTensorHolder input, es::EsTensorHolder mean, es::EsTensorHolder variance, EsGraphBuilder &graph_builder) {
  auto scale = graph_builder.CreateConst(std::vector<float>{1.0f, 1.0f, 1.0f}, std::vector<int64_t>{3});
  auto offset = graph_builder.CreateConst(std::vector<float>{0.0f, 0.0f, 0.0f}, std::vector<int64_t>{3});
  auto batchnorm = BatchNorm(input, scale, offset, mean, variance, 1e-4f, "NCHW", false);
  return batchnorm.y;
}
}
namespace es_showcase {

int RunGraph(ge::Graph &graph, const std::vector<ge::Tensor> &inputs,
             const std::string &output_prefix) {
  ge::Utils::PrintTensorsToFile(inputs, "input");
  std::map<ge::AscendString, ge::AscendString> options;
  auto *s = new (std::nothrow) ge::Session(options);
  if (s == nullptr) {
    std::cout << "Global session not ready" << std::endl;
    return -1;
  }
  static uint32_t next = 0;
  const uint32_t graph_id = next++;
  auto ret = s->AddGraph(graph_id, graph);
  if (ret != ge::SUCCESS) {
    std::cout << "AddGraph failed" << std::endl;
    delete s;
    return -1;
  }
  std::vector<ge::Tensor> outputs;
  ret = s->RunGraph(graph_id, inputs, outputs);
  if (ret != ge::SUCCESS) {
    std::cout << "RunGraph failed" << std::endl;
    (void)s->RemoveGraph(graph_id);
    delete s;
    return -1;
  }
  (void)s->RemoveGraph(graph_id);
  ge::Utils::PrintTensorsToFile(outputs, output_prefix);
  delete s;
  return 0;
}

void MakeBatchNormGraphByEsAndDump() {
  std::unique_ptr<ge::Graph> graph = MakeBatchNormGraphByEs();
  graph->DumpToFile(ge::Graph::DumpFormat::kOnnx, ge::AscendString("make_batchnorm_garph"));
}

std::unique_ptr<ge::Graph> MakeBatchNormGraphByEs() {
  // 1、创建图构建器
  auto graph_builder = std::make_unique<EsGraphBuilder>("MakeBatchNormGarph");
  // 2、创建输入节点
  auto input = graph_builder->CreateInput(0, "input", ge::DT_FLOAT, ge::FORMAT_NCHW, {1, 3, 1, 2});
  auto mean = graph_builder->CreateInput(1, "mean", ge::DT_FLOAT, ge::FORMAT_ND, {3});
  auto variance = graph_builder->CreateInput(2, "variance", ge::DT_FLOAT, ge::FORMAT_ND, {3});
  auto result = MakeBatchNormGraph(input, mean, variance, *graph_builder);
  // 3、设置输出
  (void) graph_builder->SetOutput(result, 0);
  // 4、构建图
  auto graph = graph_builder->BuildAndReset();
  return graph;
}

int MakeBatchNormGraphByEsAndRun() {
  std::unique_ptr<ge::Graph> graph = MakeBatchNormGraphByEs();
  std::vector<ge::Tensor> inputs;

  // 准备输入数据 
  std::vector<float> input_data = {1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f};
  std::vector<float> mean_data = {2.0f, 4.0f, 6.0f};
  std::vector<float> variance_data = {0.5f, 0.5f, 0.5f};

  // 创建输入tensor
  auto input_tensor = ge::Utils::StubTensor<float>(input_data, {1, 3, 1, 2}, ge::FORMAT_NCHW);
  auto mean_tensor = ge::Utils::StubTensor<float>(mean_data, {3});
  auto variance_tensor = ge::Utils::StubTensor<float>(variance_data, {3});
  inputs.push_back(*input_tensor);
  inputs.push_back(*mean_tensor);
  inputs.push_back(*variance_tensor);

  return RunGraph(*graph, inputs, "BatchNorm");
}

}  // namespace es_showcase
